### PR TITLE
Define Sequential module.

### LINKED ISSUE #1155: Implement a Sequential Module
Users often ask for this so it would be good to just add it. I can be as simple as this:

```python
class Sequential(nn.Module):
  layers: Sequence[nn.Module]

  def __call__(self, x):
    for layer in self.layers:
      x = layer(x)
    return x
```
Example usage:

```
class Foo(nn.Module):
  feature_sizes: List[int]
  @nn.compact
  def __call__(self, x):
    return Sequential([nn.Dense(sz, name=f'layers_{idx}')
                       for idx,sz in enumerate(self.feature_sizes)])(x)
```

### PR DESCRIPTION
**Pull request by melissatan**
# Adds a Sequential module and some unit tests.

Fixes #1155

## Checklist
- [ ] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).
- [x] This change is discussed in a Github issue/
      [discussion on an earlier PR](https://github.com/google/flax/pull/1156) 
- [x] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [x] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

File: flax/linen/__init__.py [hunk 1]
```diff
@@ -17,22 +17,26 @@
 
 # pylint: disable=g-multiple-import
 # re-export commonly used modules and functions
-from .activation import (celu, elu, gelu, glu, leaky_relu, log_sigmoid,
-                         log_softmax, relu, sigmoid, soft_sign, softmax,
-                         softplus, swish, silu, tanh, PReLU)
+from .activation import (PReLU, celu, elu, gelu, glu, leaky_relu, log_sigmoid,
```
Comment (melissatan) [inline flax/linen/__init__.py:20]: these changes were the result of running a formatter. i figured it's probably easier to send through this change, rather than the linter complaining.

**Review by melissatan — COMMENTED** [review]
This "Merge branch 'main'..." commit got created by Github when I resolved a merge conflict on the web UI, I tried to squash it on the command line in IDE but this commit didn't show up when I did `git rebase -i HEAD~2`. Would appreciate a hint for how to squash it. Thanks!

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Wow, you're lightning fast, great stuff!

File: flax/linen/combinators.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+"""Combinators of modules, such as a Sequential."""
+
+from typing import Callable, Sequence
+
+from flax.linen import module
+
+
+class Sequential(module.Module):
+  """Applies a linear chain of callables.
```
[the remark this answers is not in the document]
Comment (melissatan) [reply flax/linen/combinators.py:8]: Done.

[the remark this answers is not in the document]
Comment (melissatan) [reply flax/linen/combinators.py:8]: Done.

File: flax/linen/combinators.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+"""Combinators of modules, such as a Sequential."""
+
+from typing import Callable, Sequence
+
+from flax.linen import module
```
[the remark this answers is not in the document]
Comment (melissatan) [reply flax/linen/combinators.py]: Done.

File: flax/linen/combinators.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    @nn.compact
+    def __call__(self, x):
+      return Sequential([nn.Dense(layer_size, name=f'layers_{idx}')
+                         for idx, layer_size
+                         in enumerate(self.feature_sizes)])(x)
+  ```
+  """
+  layers: Sequence[Callable]
+
+  def __call__(self, x):
```
[the remark this answers is not in the document]
Comment (melissatan) [reply flax/linen/combinators.py]: Done.

File: flax/linen/combinators.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  Modules will be applied in the order that they are passed in the constructor.
+
+  The apply() method of Sequential accepts any input and forwards it to the
+  first module it contains. It chains the output sequentially to the input of
+  the next module and returns the output of the final module.
+
+  Example usage:
+
+    class Foo(nn.Module):
+        feature_sizes: List[int]
```
Comment (marcvanzee) [inline flax/linen/combinators.py:22]: Nit: for annotating arguments abstract collection types are preferred (see [here](https://docs.python.org/3/library/typing.html#typing.List)]

```suggestion
        feature_sizes: Sequence[int]
```

File: flax/linen/combinators.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      if not callable(layer):
+        raise ValueError(
+            f"Each layer should implement `__call__`, but layer `{layer}` is "
+            "not callable."
+        )
+      if idx == 0:
+        outputs = layer(outputs, *args, **kwargs)
+      else:
+        outputs = layer(outputs)
+    return outputs
```
Comment (marcvanzee) [inline flax/linen/combinators.py:44]: Maybe we can get rid of `inputs` all together, since we can assume it will be contained in `*args`? Also, I wonder whether it is correct to return the input if the module is empty... maybe we should throw an error in that case? Otherwise it could potentially introduce bugs for the users (I don't think anyone would like to use an empty Sequential module).

How about something like this? We then don't explicitely check for `callable`, but that should be covered by the type checker. I believe the original PR had that check in as an attempt to stop the type checker to complain when using `Sequence[Module]`, but we are not using that anymore here.

```suggestion
  def __call__(self, *args, **kwargs):
    if not self.layers:
      raise ValueError(f'Empty Sequential module {self.name}.')

    y = self.layers[0](*args, **kwargs)
    for layer in self.layers[1:]:
      y = layer(y)
    return y
```

File: tests/linen/linen_combinators_test.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
+"""Tests for flax.linen.combinators."""
```
Comment (marcvanzee) [inline tests/linen/linen_combinators_test.py:1]: These tests look really great!

**Review by marcvanzee — CHANGES_REQUESTED** [review]

File: flax/linen/combinators.py (hunk 6 above)
Comment (melissatan) [reply flax/linen/combinators.py:44]: Done.

File: tests/linen/linen_combinators_test.py (hunk 7 above)
Comment (melissatan) [reply tests/linen/linen_combinators_test.py:1]: Thanks for taking a look! Updated to add test for the empty module.

**Review by marcvanzee — APPROVED** [review]
Awesome, looks really good! One last question: could you add an entry to `CHANGELOG.md` for this change?

File: flax/linen/combinators.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  the next module and returns the output of the final module.
+
+  Example usage::
+
+    class Foo(nn.Module):
+        feature_sizes: Sequence[int]
+
+        @nn.compact
+        def __call__(self, x):
+        return nn.Sequential([nn.Dense(layer_size, name=f'layers_{idx}')
```
Comment (avital) [inline flax/linen/combinators.py:26]: Can we make the example include a non-linearity like `relu`? That's going to be necessary for pretty much all use-cases of `nn.Sequential` and people will want to see how it's done.

Comment (melissatan) [reply flax/linen/combinators.py:26]: Ok, doing this in #1978 .
